Sources/SwiftCompilerPluginMessageHandling/JSON/JSONDecoding.swift

```diff
@@ -0,0 +1,1286 @@
+  }
+
+  static func parseFloatingPoint<Floating: BinaryFloatingPoint>(source: UnsafeBufferPointer<UInt8>) -> Floating? {
+    var endPtr: UnsafeMutablePointer<CChar>? = nil
+    let value: Floating?
+    if (Floating.self == Double.self) {
+      value = Floating(exactly: strtod(source.baseAddress!, &endPtr))
+    } else if (Floating.self == Float.self) {
+      value = Floating(exactly: strtof(source.baseAddress!, &endPtr))
+    } else {
+      fatalError("unsupported floating point type")
+    }
+    guard endPtr! == source.baseAddress! + source.count else {
+      return nil
+    }
+    return value
+  }
+
+  static func parseHexIntegerDigits<Integer: FixedWidthInteger>(source: UnsafeBufferPointer<UInt8>) -> Integer? {
+    var source = source[...]
+    var value: Integer = 0
+    var overflowed: Bool = false
+    while let digit = source.popFirst() {
+      let digitValue: Integer
+      switch digit {
+      case UInt8(ascii: "0")...UInt8(ascii: "9"):
+        digitValue = Integer(truncatingIfNeeded: digit &- UInt8(ascii: "0"))
+      case UInt8(ascii: "a")...UInt8(ascii: "f"):
+        digitValue = Integer(truncatingIfNeeded: digit &- UInt8(ascii: "a") &+ 10)
+      case UInt8(ascii: "A")...UInt8(ascii: "F"):
+        digitValue = Integer(truncatingIfNeeded: digit &- UInt8(ascii: "A") &+ 10)
+      default:
+        return nil
+      }
+      (value, overflowed) = value.multipliedReportingOverflow(by: 16)
+      guard !overflowed else {
+        return nil
+      }
+      (value, overflowed) = value.addingReportingOverflow(digitValue)
+      guard !overflowed else {
+        return nil
+      }
+    }
+    return value
+  }
+}
+
+extension JSONMapValue {
+  /// Get value buffer for .number, .string, and .simpleString.
+  @inline(__always)
+  func valueBuffer() -> UnsafeBufferPointer<UInt8> {
+    UnsafeBufferPointer<UInt8>(
+      start: UnsafePointer(bitPattern: data[1]),
+      count: data[2]
+    )
+  }
+
+  @inline(__always)
+  func asString() -> String? {
+    if self.is(.asciiSimpleString) || self.is(.simpleString) {
+      return _JSONStringParser.decodeSimpleString(source: valueBuffer())
+    }
+    if self.is(.string) {
+      return _JSONStringParser.decodeStringWithEscapes(source: valueBuffer())
+    }
+    return nil
+  }
+
+  /// Returns true if this value represents a string and equals to 'str'.
+  ///
+  /// This is faster than 'value.asString() == str' because this doesn't
+  /// instantiate 'Swift.String' unless there are escaped characters.
+  func equals(to str: String) -> Bool {
+    if self.is(.asciiSimpleString) {
+      let lhs = valueBuffer()
+      var str = str
+      return str.withUTF8 { rhs in
+        if lhs.count != rhs.count {
+          return false
+        }
+        guard let lBase = lhs.baseAddress, let rBase = rhs.baseAddress else {
+          // If either `baseAddress` is `nil`, both are empty so returns `true`.
+          return true
+        }
+        return memcmp(lBase, rBase, lhs.count) == 0
```

Comment:
I don't think it's worth it. This is small enough, and I really do not want to cross modules for this.

Reply:
Because we couldn't use `Darwin`.  We used to use `Collection.elementsEqual(_:)`

Reply:
Ah sorry, I mistakingly thought you were talking about this entire block
```
        if lhs.count != rhs.count {
          return false
        }
        guard let lBase = lhs.baseAddress, let rBase = rhs.baseAddress else {
          // If either `baseAddress` is `nil`, both are empty so returns `true`.
          return true
        }
        return memcmp(lBase, rBase, lhs.count) == 0
```
But actually it was `compareMemroy(_:_:_:)` specifically.
Yeah, it was only because we couldn't use `Darwin`, and we didn't support Windows (I.e. `ucrt`). We can use `memcmp` directly in SwiftSyntax as well now.